### ISSUE TITLE
Unconditionally adds Requires.private to .pc files even if not actually required

### ISSUE THREAD

**Issue by ximion** [mesonbuild/meson#3970]
Hi!

Since a few releases, Meson unconditionally adds `Requires.private` libraries to pkg-config .pc files, even if none of the interfaces of these libraries are actually exposed at all in the shared library that is to be installed.
For example, I have a library that uses libxml2 internally, but none if its types are actually exposed and nobody using my library needs to have the development files for libxml2 actually installed on their system.

However, when they now want to link against my library, pkg-config complains with 
```
pkg-config --cflags appstream
Package libxml-2.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `libxml-2.0.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libxml-2.0', required by 'appstream', not found
```
This behavior change broke quite a few things, and the only way I seem to be able to control what ends up in the .pc files is to postprocess it with sed, which seems suboptimal.

Can support for controlling what ends up in `Requires.private` & Co. be added, please?

**Comment by ximion** [comment]
This *might* actually be a bug in pkg-config itself, because as far as I understood the documentation, `Libs.private` and `Requires.private` only matter for static linking. So if I don't do static linking, pkg-config shouldn't actually search for .pc files in `Requires.private`.
Regardless, it would be nice if Meson had some way of controlling this behavior.

**Comment by xclaesse** [comment]
Yes, you need libxml2 to static link on your library. Note that if you're not going to provide static version of your library you can use shared_library() instead of library() and it won't add requires.private in PC file in that case.

**Comment by ignatenkobrain** [comment]
this is really bug in pkg-config. And I believe it has been fixed in [pkgconf](https://github.com/pkgconf/pkgconf)

**Comment by ximion** [comment]
@ignatenkobrain That would make sense, actually @xclaesse both replies make sense. Potentially Debian should switch to pkgconf as well, I might investigate that possibility.
I wonder whether Meson should not write the `Requires.private` line even for `library()` targets though if Meson is run with `--default-library=shared` (the default option) - that would yield a result that is more predictable (at least for me).

**Comment by nirbheek** [comment]
> I wonder whether Meson should not write the Requires.private line even for library() targets though if Meson is run with --default-library=shared (the default option)

I thought that was already the case, if not we should fix it.

**Comment by ximion** [comment]
> I thought that was already the case, if not we should fix it.

I just tested this, it definitely doesn't happen here (v0.47.1). Changing this behavior would definitely fix this issue for me :-)

**Comment by xclaesse** [comment]
I think it's weird to have a different .pc file depending on that build option. Developers won't usually pass --default-library=both and they will wonder why their pc files are wrong.

Note that `pkg-config --cflags` also include cflags from private libraries, and that's on purpose. There is a long discussion about this in a pkg-config bug: https://bugs.freedesktop.org/show_bug.cgi?id=105572. I made a patch there to add `Requires.internal`, but it never got reviewed.

Unless you never intended to provide a working static library, then you definitely should use shared_library() instead of library().

**Comment by ximion** [comment]
> I think it's weird to have a different .pc file depending on that build option. Developers won't usually pass --default-library=both and they will wonder why their pc files are wrong.

But they aren't wrong - if the developer wants shared libraries, they get shared libraries with the .pc file appropriate for them.

> Note that pkg-config --cflags also include cflags from private libraries, and that's on purpose. 

That's really dumb though - it leads to overlinking of libraries (which will later get stripped out again anyway when built with `--as-needed`) and increases the dependencies of a package for absolutely no reason at all for most of these cases.
I think Simon's post on the linked bug report sums up the conundrum perfectly. I think we either need `Requires.internal` as you suggested, or at least have Meson give the user a bit of control over what ends up in the .pc file for invisible dependencies, for people who know what they want (my library is explicitly designed in a way that makes all other libs it links to an internal implementation detail).

**Comment by xclaesse** [comment]
> That's really dumb though - it leads to overlinking of libraries

No, the point here is to avoid overlinking, it won't link on stuff from Requires.private unless you do static link.

I don't get why it's such a big deal, the only implication is you have a few more build dependencies to install to build your app. It has no impact on the resulting application binary.

Note that even if you build only the shared library, you still need `Requires.private` for *visible dependencies* (as described in Simon's post), and that would be not be backward compatible change.

> Do you have a link to your pkg-config patch?

There are attached on the bug I linked above.

**Comment by xclaesse** [comment]
> That would be the expected result, but in reality that's not what pkg-config does - it simply includes all dependencies, no matter whether I link statically or dynamically, and no matter whether they are private or not.

Just to be clear, `pkg-config --cflags` will add cflags from Requires.private, but `pkg-config --libs` will NOT add linker flags from Requires.private. So there is no overlinking. The reason is Requires.private is for the case were your library exposes that dependency in its headers but an app using your library does not usually need to link on that dependency (aka Simon's *visible dependencies*). There is no way to express in a .pc file an *invisible dependency*, that's a limitation of the format.

> Pointless dependencies make transitions harder and builds slower due to additional stuff needing to be installed. It's better to avoid that if possible. Additionally, the change in Meson actually was the thing that broke stuff for us and made me discover this issue via automatic tests.

Who is "us" ? If you mean debian, they have a policy to build both shared and static libraries in their packages AFAIK, so those dependencies are required. If you were missing them previously, that means your .pc has always been wrong.

**Comment by ximion** [comment]
> Just to be clear, pkg-config --cflags will add cflags from Requires.private, but pkg-config --libs will NOT add linker flags from Requires.private.

Indeed, I confused myself here. The erros is actually something like `Package 'libxml-2.0', required by 'mypackage', not found`. If you would *just* add the pkg-config file of the library, the build will pass without any errors, which is actually quite dumb.

> Who is "us" ? If you mean debian, they have a policy to build both shared and static libraries in their packages AFAIK

No, we don't. We *have* to build the shared library unless there is a good reason not to, and the static library is only built on special occasions. It often is omitted to avoid anything in the distribution being statically linked by accident, but in cases where it is genuinely useful to have static libraries, they are included. This is a rather rare thing though (I did a quick and very inaccurate check (contains false positives for static libs), and of all libraries in Debian a maximum of 34.5% have static libs).

> If you were missing them previously, that means your .pc has always been wrong.

It was definitely fine before, only a shared library is shipped. Adjusting Meson to create a smaller .pc file when only a shared library is built indeed feels like a good way to go here, especially if this was the original intent anyway.
Extending pkg-config to include internal requires as separate field would be the ultimate fix, but that will take more time and work.

**Comment by kalev** [comment]
WIth my Fedora packager hat on, I'd like to see this changed as well, just like @ximion described :)

**Comment by Ericson2314** [comment]
https://bugs.freedesktop.org/show_bug.cgi?id=105572 which was linked is now https://gitlab.freedesktop.org/pkg-config/pkg-config/-/issues/7. It seemed like there was agreement but I don't think anything was merged, so I left a comment asking what happened.

**Comment by Apteryks** [comment]
> this is really bug in pkg-config. And I believe it has been fixed in [pkgconf](https://github.com/pkgconf/pkgconf)

It's not, unfortunately.  Requires.private packages must be consulted also just to find relevant CFLAGS for example, see https://github.com/pkgconf/pkgconf/commit/2d26db3a984e7a69cedc6274c22119de4bcf1b40#diff-9f79b045541879ce71b26692de67a5b727a532719229fdaa2635ccbf64863fa1R132

Also forcing everyone onto a different pkg-config implementation is also suboptimal, if the root cause really is Meson adding these Requires.private for no benefit.
